Add support for `rubicon-objc`

timrid opened the pull request:
This adds support for [`rubicon-objc`](https://github.com/beeware/rubicon-objc) to potentially support iOS. Currently I only tested this on macOS and not on iOS.

On macOS it is possible to activate the `rubicon-objc` framework by setting the environment variable `BLEAK_COREBLUETOOTH_FRAMEWORK` to `rubicon-objc`.

I just wanted to share this as a first draft until I actual did tests with an iPhone.

Fixes #1833

dlech commented:
I was hoping for something a bit less invasive to the existing code, e.g. creating a `_pyobjc_compat/` folder with modules having the same name and containing classes that have the same API as PyObjC that calls into the equivalent Rubicon API.

Then the existing code would just have to be modified like:

```python
try:
    from CoreBluetooth import CBUUID, ...
except ImportError:
    from ._pyobjc_compat.CoreBluetooth import CBUUID, ...
```

dlech commented:
> Maybe I could split this PR in smaller chunks to simplify the review? Would that be a compromise?

Smaller chunks is always helpful.

I think the only thing here that I will really strongly object to is changing the files in `typings` to be for anything other than PyObjC (fixes are OK, of course). 

Rubicon ObjC should have it's own separate typings. I was using it for USB stuff recently and so I have some idea of how I would want to do this, so I don't mind helping with that part.

timrid commented:
I have rebased the PR to the current version and built it based on https://github.com/hbldh/bleak/pull/1922. You can use
`tox -- --bleak-hci-transport=serial:/dev/tty.usbmodem1401` to test both `pyobjc` and `rubicon-objc` completely. This means unit tests and type checks for both versions. It takes a while, but then you can be sure that everything still works.

I would adapt the documentation afterwards, once the details have been clarified. I also removed everything specific to iOS. I will do that later via a separate PR.

johnzhou721 commented on bleak/backends/corebluetooth/_objc_compat.py:
```diff
@@ -0,0 +1,322 @@
+            raise TypeError(f"Expected int, got {type(ret)}")
+        return ret
+
+    def to_str(objc_obj: NSString) -> str:
+        """Converts an NSString to a Python string."""
+        return py_from_ns(objc_obj)
```
strs autoconverts in rubicon-objc as well.  So this shim is not needed.  Is this for consistency?

johnzhou721 commented on bleak/backends/corebluetooth/_objc_compat.py:
```diff
@@ -0,0 +1,322 @@
+        # Normally rubicon-objc parses the type hints from `func` for e.g. ctypes types, but the type hints
+        # we use are not compatible. So we remove all type hints with `no_type_check` before passing the
+        # function to rubicon-objc.
+        return cast(F, _objc_method(no_type_check(func)))
```
If we don't have type-check, return values, etc. are going to be very inconsistently handled.  All return values are None, and all parameters are of type `id`.

Does there exist any other functions other than this type of signatures in the codebase?  If so, I think it's not too hard to convert the type hints used in the codebase to actual ctypes type hints.
